### PR TITLE
Race condition in concurrent shutdown/startup (1: retry)

### PR DESCRIPTION
**Pull request by hkdsun**
This is the retry method that waits for steady state
@fw42

### CODE REVIEW

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 1]
```diff
@@ -84,8 +84,16 @@ def schedule
       end
 
       # reloads the schedule from redis
-      def reload_schedule!
+      def reload_schedule!(verify_with: nil, max_retries: 7)
+        retries = 0
+
         @schedule = all_schedules
+        while verify_with && retries < max_retries &&
+              !(@schedule && @schedule.eql?(verify_with))
```
Comment (fw42) [inline lib/resque/scheduler/scheduling_extensions.rb]: do we really care about equality here? isn't it enough to check that some schedule exists?

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 2]
```diff
@@ -84,8 +84,16 @@ def schedule
...
+          @schedule = all_schedules
+          retries += 1
+          sleep 2**(retries - rand + 1) * 0.1 unless @schedule
```
Comment (fw42) [inline lib/resque/scheduler/scheduling_extensions.rb]: why the `rand` stuff?

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 3]
```diff
@@ -84,8 +84,16 @@ def schedule
...
+        end
```
Comment (fw42) [inline lib/resque/scheduler/scheduling_extensions.rb]: I think I would rewrite this method a bit for clarity.. And I don't like `while` loops too much :-)
```ruby
def reload_schedule(verify_with: nil, max_retries: 7)
  (1..max_retries).each do |tries|
    schedule = all_schedules

    if schedule.eql?(verify_with)
      return @schedule = schedule
    end

    sleep 0.1 * (2**tries)
  end
end
```

File: lib/resque/scheduler/scheduling_extensions.rb [hunk 4]
```diff
@@ -101,8 +109,10 @@ def all_schedules
 
       # clean the schedules as it exists in redis, useful for first setup?
       def clean_schedules(keys = non_persistent_schedules)
-        keys.each do |key|
-          remove_schedule(key)
+        redis.multi do
```
Comment (fw42) [inline lib/resque/scheduler/scheduling_extensions.rb]: why this? what problem does this line here solve? is that necessary? how is it related to the rest of this PR?

**Comment by fw42** [comment]
I guess this PR doesn't actually "fix" the race condition, it just makes it less likely, right? The same problem could still happen with enough servers and some bad luck.

File: test/multi_process_test.rb [hunk 5]
```diff
@@ -17,10 +17,10 @@
     end
 
     processes.times do |n|
-      threads << Thread.new do
-        sleep n * 0.1
+      pids << fork_with_marshalled_pipe_and_result do
+        sleep 1.0 / (n + 1) * 0.2
```
Comment (fw42) [inline test/multi_process_test.rb]: where does this come from? why is it this exact value?

File: test/multi_process_test.rb [hunk 6]
```diff
@@ -28,10 +28,90 @@
...
 
-    threads.each(&:join)
+    counts += get_results_from_children(pids)
 
     counts.each_with_index do |c, i|
       assert_equal schedule_count, c, "schedule count is incorrect (c: #{i})"
     end
   end
+
+  test 'focus concurrent shutdowns and startups do not corrupt the schedule' do
```
Comment (fw42) [inline test/multi_process_test.rb]: I don't understand this sentence.. what does "focus" mean?

Comment (fw42) [inline test/multi_process_test.rb]: this test fails on master?

File: lib/resque/scheduler/scheduling_extensions.rb (hunk 2 above)
Comment (hkdsun) [inline lib/resque/scheduler/scheduling_extensions.rb]: so that two or more processes don't back off in a uniform manner, again colliding with each other

Comment (hkdsun) [inline lib/resque/scheduler/scheduling_extensions.rb]: i suppose that might not matter in this case if we're just reading from redis using `all_schedules`

File: test/multi_process_test.rb (hunk 6 above)
Comment (hkdsun) [inline test/multi_process_test.rb]: focus should go away..

and yes it fails. every time

File: lib/resque/scheduler/scheduling_extensions.rb (hunk 4 above)
Comment (hkdsun) [inline lib/resque/scheduler/scheduling_extensions.rb]: Cleaning schedules must be done in a MULTI transaction every time.

This method is called two times in `resque-scheduler`: once during every instance's shutdown outside a transaction which causes schedules being loaded partially (e.g. if `reload_schedule!` is  called while an instance is shutting down, then we get a partial schedule) and the other time is inside a transaction when an instance is starting up

**Comment by hkdsun** [comment]
>I guess this PR doesn't actually "fix" the race condition, it just makes it less likely, right?

Right. And it could let us know when it has happened

File: lib/resque/scheduler/scheduling_extensions.rb (hunk 4 above)
Comment (fw42) [inline lib/resque/scheduler/scheduling_extensions.rb]: the tests break if you remove this here?

**Comment by hkdsun** [comment]
closing these for now

**Pull request closed without merging**